builtin/audit/socket/backend.go

```diff
@@ -0,0 +1,167 @@
+package socket
+
+import (
+	"bytes"
+	"fmt"
+	"net"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/vault/audit"
+	"github.com/hashicorp/vault/helper/duration"
+	"github.com/hashicorp/vault/logical"
+)
+
+func Factory(conf *audit.BackendConfig) (audit.Backend, error) {
+	if conf.Salt == nil {
+		return nil, fmt.Errorf("nil salt passed in")
+	}
+
+	address, ok := conf.Config["address"]
+	if !ok {
+		return nil, fmt.Errorf("address is required")
+	}
+
+	socketType, ok := conf.Config["socket_type"]
+	if !ok {
+		socketType = "tcp"
+	}
+
+	writeDeadline, ok := conf.Config["write_deadline"]
+	if !ok {
+		writeDeadline = "2s"
+	}
+	writeDuration, err := duration.ParseDurationSecond(writeDeadline)
+	if err != nil {
+		return nil, err
+	}
+
+	format, ok := conf.Config["format"]
+	if !ok {
+		format = "json"
+	}
+	switch format {
+	case "json", "jsonx":
+	default:
+		return nil, fmt.Errorf("unknown format type %s", format)
+	}
+
+	// Check if hashing of accessor is disabled
+	hmacAccessor := true
+	if hmacAccessorRaw, ok := conf.Config["hmac_accessor"]; ok {
+		value, err := strconv.ParseBool(hmacAccessorRaw)
+		if err != nil {
+			return nil, err
+		}
+		hmacAccessor = value
+	}
+
+	// Check if raw logging is enabled
+	logRaw := false
+	if raw, ok := conf.Config["log_raw"]; ok {
+		b, err := strconv.ParseBool(raw)
+		if err != nil {
+			return nil, err
+		}
+		logRaw = b
+	}
+
+	conn, err := net.Dial(socketType, address)
+	if err != nil {
+		return nil, err
+	}
+
+	b := &Backend{
+		connection: conn,
+		formatConfig: audit.FormatterConfig{
+			Raw:          logRaw,
+			Salt:         conf.Salt,
+			HMACAccessor: hmacAccessor,
+		},
+		writeDuration: writeDuration,
+		address:       address,
+		socketType:    socketType,
+	}
+
+	switch format {
+	case "json":
+		b.formatter.AuditFormatWriter = &audit.JSONFormatWriter{}
+	case "jsonx":
+		b.formatter.AuditFormatWriter = &audit.JSONxFormatWriter{}
+	}
+
+	return b, nil
+}
+
+// Backend is the audit backend for the socket audit transport.
+type Backend struct {
+	connection net.Conn
+
+	formatter    audit.AuditFormatter
+	formatConfig audit.FormatterConfig
+
+	writeDuration time.Duration
+	address       string
+	socketType    string
+
+	sync.Mutex
+}
+
+func (b *Backend) GetHash(data string) string {
+	return audit.HashString(b.formatConfig.Salt, data)
+}
+
+func (b *Backend) LogRequest(auth *logical.Auth, req *logical.Request, outerErr error) error {
+	var buf bytes.Buffer
+	if err := b.formatter.FormatRequest(&buf, b.formatConfig, auth, req, outerErr); err != nil {
+		return err
+	}
+
+	b.Lock()
+
+	b.connection.SetDeadline(time.Now().Add(b.writeDuration))
+	_, err := b.connection.Write(buf.Bytes())
+
+	b.Unlock()
+	if err != nil {
+		b.Reload()
```

Comment:
I think we should remove the locking from Reload and instead call it from within the locked section in this and LogResponse. The reason is that if there is an error and we want to reset the connection, we don't want to give up the lock and possibly have other writers grab it and attempt to write to the bad connection...we want to reset the connection ASAP.